src/topotoolbox/stream_object.py

```diff
@@ -0,0 +1,78 @@
+"""This module contains the StreamObject class.
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+
+from .flow_object import FlowObject
+
+# pylint: disable=no-name-in-module
+from . import _flow  # type: ignore
+
+_all_ = ['StreamObject']
+
+
+class StreamObject():
+    """A class to represent stream flow accumulation based on a FlowObject.
+    """
+
+    def __init__(self, flow: FlowObject) -> None:
+
+        acc = np.zeros_like(flow.target, order='F', dtype=np.float32)
+        weights = np.ones_like(flow.target, order='F', dtype=np.float32)
+
+        _flow.flow_accumulation(
+            acc, flow.source, flow.direction, weights, flow.shape)
+
+        self.path = flow.path
+        self.name = flow.name
+
+        # raster metadata
+        self.acc = acc
+        self.shape = flow.shape
+
+        # georeference
+        self.bounds = flow.bounds
+        self.transform = flow.transform
+        self.crs = flow.crs
```

Comment:
The `STREAMobj` defined in MATLAB is a so-called "node attribute list" that is an array of pixel indices that are considered part of the stream network. Typically those pixels are ones with a flow accumulation value greater than some threshold. The constructor for `STREAMobj` looks something like

```
STREAMobj(FD : FLOWobj, threshold, units)
```

where `threshold` can be either a number, in which case the stream network is the collection of pixels with a flow accumulation greater than that threshold when measured in the supplied `units` -- `flow_accumulation` returns the flow accumulation in pixels, but you could convert that to square meters or square kilometers using the `cellsize`.

`threshold` could also be a Boolean-valued `GRIDobj`, in which case the stream pixels are those where `threshold` is True.

I think we should keep the same or a similar interface for `StreamObject` here. What you have currently is a `GridObject` by another name.

Reply:
We also need to hang on to the connectivity information from the `FlowObject` so that we can traverse the stream network and do various calculations with it. MATLAB doesn't store the entire `source` and `target` arrays, but only the subsets that are part of the stream network. Find all the entries in `source` that are labeled `True` in `threshold` and then store those entries and the corresponding `target` and `direction` entries in the `StreamObject`.

Let me know if that makes sense -- I am finding it very confusing at the moment.

Reply:
I think I understand. So `threshold` is either a `int | float` or a `GridObject`. 

1. If it's a number, we'll check if a cell is `>= threshold`. If so, it'll be part of the stream.
2. If it's a (bool) `GridObject` only the cells where the value is 1 will be part of the stream.

Only the indexes  of the stream are going to be saved in the StreamObject.
![grafik](https://github.com/user-attachments/assets/2e622b93-09e3-45a0-93fb-df2d1532d24d)
So in this case, where the blue line is the resulting stream, the array would contain the indexes like this: `[1,2,5,8]` correct?

So the `source`, `target` and `direction` arrays are also just arrays that contain the respective values when parsing through the  `[1,2,5,8]` array?

So if I continue the example (but filled with example values instead of indexes this time):
![grafik](https://github.com/user-attachments/assets/831f1c6d-40b8-4115-83eb-c11d14fd9ce9)
It should look like this: `source=[1,4,7,6], target=[2,1,4,5], direction=[5,5,8,6]`?

Reply:
> I think I understand. So `threshold` is either a `int | float` or a `GridObject`.
> 
>     1. If it's a number, we'll check if a cell is `>= threshold`. If so, it'll be part of the stream.
> 
>     2. If it's a (bool) `GridObject` only the cells where the value is 1 will be part of the stream.
> 

This is correct

> Only the indexes of the stream are going to be saved in the StreamObject. ![grafik](https://private-user-images.githubusercontent.com/69400012/366502545-2e622b93-09e3-45a0-93fb-df2d1532d24d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxMjE5MTYsIm5iZiI6MTcyNjEyMTYxNiwicGF0aCI6Ii82OTQwMDAxMi8zNjY1MDI1NDUtMmU2MjJiOTMtMDllMy00NWEwLTkzZmItZGYyZDE1MzJkMjRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDA2MTMzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI1MmMzZmMwZGM5YmQyNTM0OWRjZmRmNmQ5ZTQyY2Q3NWMzMmU3Mjc5M2VmNTc3MzBjMjAyNGI0MmJkNTU1OWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1HRe_ugbfbh_nP4px-uqX7wnVJpBkXvp2xHiNBXFL1c) So in this case, where the blue line is the resulting stream, the array would contain the indexes like this: `[1,2,5,8]` correct?

Yes, this is what we need.

> So the `source`, `target` and `direction` arrays are also just arrays that contain the respective values when parsing through the `[1,2,5,8]` array?
> 
> So if I continue the example (but filled with example values instead of indexes this time): ![grafik](https://private-user-images.githubusercontent.com/69400012/366507382-831f1c6d-40b8-4115-83eb-c11d14fd9ce9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxMjE5MTYsIm5iZiI6MTcyNjEyMTYxNiwicGF0aCI6Ii82OTQwMDAxMi8zNjY1MDczODItODMxZjFjNmQtNDBiOC00MTE1LTgzZWItYzExZDE0ZmQ5Y2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDA2MTMzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIyZDM2MGRmNzhkZWZkYTEyZTcyNmVmYzZiMzlhYWNmNTUzMTIxOWUxYTA2NTg1MDhlNDJhYWFkNGVkZjk1ZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PIfLOjUwxHiYphw3fidmVbaM3wL6c10DA2OU1G9PidY) It should look like this: `source=[1,4,7,6], target=[2,1,4,5], direction=[5,5,8,6]`?

The `source`, `target` and `direction` arrays are "edge attribute lists" while the `stream` array is a "node attribute list." If you think of the flow network as a graph, the pixels of the raster are the nodes and edges connect two pixels. Edge attribute lists label the edges, and node attribute lists label the nodes. For now we only route flow in a single direction per pixel, so there is one edge for each node, and the lists are approximately the same length. The only potential difference depends on how you treat nodes that have no downstream neighbor (i.e. outlets or sinks). In the future we will want to support multiple flow directions, in which case there can be more edges than nodes.

In your example, if we assume that the stream flows in the direction 1 -> 2 -> 5 -> 8, then your three arrays will be

- `source = [1,2,5,8]`
- `target = [2,5,8,-1]` (the flow routing routines return a target of `-1` for any pixel that does not have a downstream neighbor).
- `direction = [4,1,1,0]` (the `direction` array encodes direction as a bitfield. These may not be the right values, because it depends on the memory order of the array).

The MATLAB implementation leaves off the `8 -> -1` edge, so it will have `source = [1,2,5]` and `target = [2,5,8]`, so you could also go through and filter out edges that have `target = -1` if you wanted to.